**Bowtie 2** is a fast and efficient aligner for aligning reads to reference sequences, particularly for input reads of 50-1000 characters and relatively long (e.g. mammalian) reference genomes. For more information, see the tool's [Sourceforge page](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml) and [GitHub repo](https://github.com/BenLangmead/bowtie2).

Function Call
=============

```python
tc.bowtie2(
  	inputs=,
  	output_path=None,
  	tool_args="",
  	database_name="standard",
  	database_version="1",
  	is_async=False,
)
```

Function Arguments
------------------

| Argument           | Use in place of:    | Description                                                                                                                                                                                                                                              |
| :----------------- | :------------------ | :------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `inputs`           | `-U`                | Path to one or more files to use as input. The files can be a local or remote, see [Using Files](../../getting-started/using-files.md).                                                                                          |
| `output_path`      | `-S`                | (optional) Path (directory) to where the output files will be downloaded. If omitted, skips download. The files can be a local or remote, see [Using Files](../../getting-started/using-files.md).                               |
| `tool_args`        | all other arguments | (optional) Additional arguments to be passed to Bowtie 2. This should be a string of arguments like the command line. See [Supported Additional Arguments](#supported-additional-arguments) for more details. |
| `database_name`    | `-x`\*              | (optional) Name of database to use for Bowtie 2 alignment. Defaults to `"GRCh38_noalt_as"` (human genome).                                                                                                                                               |
| `database_version` | `-x`\*              | (optional) Version of database to use for Bowtie 2 alignment. Defaults to `"1"`.                                                                                                                                                                         |
| `is_async`         |                     | Whether to run a job asynchronously.  See [Async Runs](../../feature-reference/async-runs.md) for more.                                                                                                                                                                  |

See the [Databases](#databases) section for more details.

Output Files
------------

A Bowtie 2 run will output these files into  `output_path`:

- `bowtie2.log`: Log outputted to `stderr`.
- `bowtie2_output.sam`: SAM file outputted by Bowtie 2.

Notes
-----

### Paired-end inputs

Paired-end read inputs can be provided through `inputs`.

Tool Versions
=============

Toolchest currently supports version **2.4.4.** of Bowtie 2. 

Databases
=========

Toolchest currently supports the following databases for Bowtie 2:

| `database_name`   | `database_version` | Description                                                 |
| :---------------- | :----------------- | :---------------------------------------------------------- |
| `Ash1_v2.0`       | `1`                | Human / Ash2.0, JHU source<sup>1</sup>                      |
| `GRCh38_noalt_as` | `1`                | Human / GRCh38 no-alt analysis set, NCBI source<sup>1</sup> |
| `GRCm39`          | `1`                | Mouse / GRCm39, NCBI source<sup>1</sup>                     |

<sup>1</sup>These database indexes were generated by [the Langmead Lab at Johns Hopkins](https://langmead-lab.org/) and can be found on [the lab's database index page](https://benlangmead.github.io/aws-indexes/bowtie).

Supported Additional Arguments
==============================

- `-3`
- `-5`
- `-a`
- `--align-paired-reads`
- `-D`
- `--dovetail`
- `--dpad`
- `--end-to-end`
- `-f`
- `-F`
- `--fast`
- `--fast-local`
- `--ff`
- `--fr`
- `--gbar`
- `-i`
- `-I`
- `--ignore-quals`
- `--int-quals`
- `-k`
- `-L`
- `--local`
- `--ma`
- `--maxins`
- `--minins`
- `--mp`
- `-N`
- `--n-ceil`
- `--no-1mm-upfront`
- `--no-contain`
- `--no-discordant`
- `--nofw`
- `--no-hd`
- `--no-mixed`
- `--non-deterministic`
- `--no-overlap`
- `--norc`
- `--no-sq`
- `--no-unal`
- `--np`
- `--omit-sec-seq`
- `--phred33`
- `--phred64`
- `--preserve-tags`
- `-q`
- `--qc-filter`
- `--qseq`
- `-r`
- `-R`
- `--rdg`
- `--rf`
- `--rfg`
- `--rg`
- `--rg-id`
- `-s`
- `--sam-append-comment`
- `--sam-no-qname-trunc`
- `--score-min`
- `--seed`
- `--sensitive`
- `--sensitive-local`
- `--soft-clipped-unmapped-tlen`
- `--solexa-quals`
- `-t`
- `--tab5`
- `--tab6`
- `--time`
- `--trim-to`
- `-u`
- `--very-fast`
- `--very-fast-local`
- `--very-sensitive`
- `--very-sensitive-local`
- `-X`
- `--xeq`

Additional arguments can be specified under the `tool_args` argument.